sys/include/net/eui_provider.h

```diff
@@ -0,0 +1,209 @@
+ *
+ * @return      0 on success, next provider in eui64_conf_t will be
+ *              used otherwise.
+ *              Will fall back to @see luid_get_eui64 eventually.
+ */
+typedef int (*netdev_get_eui64_cb_t)(const void *arg, eui64_t *addr);
+
+/**
+ * @brief Structure to hold providers for EUI-48 addresses
+ */
+typedef struct {
+    netdev_get_eui48_cb_t provider; /**< function to provide an EUI-48                  */
+    const void *arg;                /**< argument to the provider function              */
+    netdev_type_t type;             /**< device type to match or `NETDEV_ANY`           */
+    uint8_t index;                  /**< device index to match or `NETDEV_INDEX_ANY`    */
+} eui48_conf_t;
+
+/**
+ * @brief Structure to hold providers for EUI-64 addresses
+ */
+typedef struct {
+    netdev_get_eui64_cb_t provider; /**< function to provide an EUI-64                  */
+    const void *arg;                /**< argument to the provider function              */
+    netdev_type_t type;             /**< device type to match or `NETDEV_ANY`           */
+    uint8_t index;                  /**< device index to match or `NETDEV_INDEX_ANY`    */
+} eui64_conf_t;
+
+/**
+ * @brief Generates an EUI-48 address for the netdev interface.
+ *
+ * @note It is possible to supply a board-specific, constant address
+ *       by implementing a EUI-48 provider function.
+ *       If no such function is available, this will fall back to
+ *       @ref luid_get_eui48.
+ *
+ * @param[in] netdev    The network device for which the address is
+ *                      generated.
+ * @param[out] addr     The generated EUI-48 address
+ *
+ */
+void netdev_eui48_get(netdev_t *netdev, eui48_t *addr);
+
+/**
+ * @brief Generates an EUI-64 address for the netdev interface.
+ *
+ * @note It is possible to supply a board-specific, constant address
+ *       by implementing a EUI-64 provider function.
+ *       If no such function is available, this will fall back to
+ *       @ref luid_get_eui64.
+ *
+ * @param[in] netdev    The network device for which the address is
+ *                      generated.
+ * @param[out] addr     The generated EUI-64 address
+ *
+ */
+void netdev_eui64_get(netdev_t *netdev, eui64_t *addr);
+
+/**
+ * @brief   Get a short unicast address from an EUI-64
+ *
+ * The resulting address is built from the provided long address.
+ * The last two bytes of the long address will be used as the short
+ * address with the first bit cleared.
+ *
+ * @param[in]  addr_long    the address to base the short address on
+ * @param[out] addr_short   memory location to copy the address into.
+ */
+static inline void eui_short_from_eui64(eui64_t *addr_long,
+                                        network_uint16_t *addr_short)
+{
+    /* https://tools.ietf.org/html/rfc4944#section-12 requires the first bit to
+     * 0 for unicast addresses */
+    addr_short->u8[0] = addr_long->uint8[6] & 0x7F;
+    addr_short->u8[1] = addr_long->uint8[7];
+}
```

Comment:
I'm realizing just now that this is not implementing the same workaround to have unique short address that is done for `luid_get_short()`.

```
    luid_base(addr, sizeof(*addr));
    addr->u8[1] ^= lastused++;

    /* https://tools.ietf.org/html/rfc4944#section-12 requires the first bit to
     * 0 for unicast addresses */
    addr->u8[0] &= 0x7F;
```

Should this call `luid_get_short` instead? OR change it to:

```
static inline void eui_short_from_eui64(eui64_t *addr_long,
                                        network_uint16_t *addr_short)
{
    /* https://tools.ietf.org/html/rfc4944#section-12 requires the first bit to
     * 0 for unicast addresses */
    addr_short->u8[0] = addr_long->uint8[0] & 0x7F;
    addr_short->u8[1] ^= addr_long->uint8[1];
}
```

Reply:
Huh? Byte 7 will change if we got the EUI64 from luid and there are multiple interfaces

```C
void luid_get_eui64(eui64_t *addr)
{
    luid_base(addr, sizeof(*addr));
    addr->uint8[7] ^= lastused++;

    eui64_set_local(addr);
    eui64_clear_group(addr);
}

void luid_netdev_get_eui64(const netdev_t *netdev, eui64_t *addr)
{
    luid_base(addr, sizeof(*addr));
#ifdef MODULE_NETDEV_REGISTER
    addr->uint8[6] ^= netdev->type;
    addr->uint8[7] ^= netdev->index;
#else
    /* we should only get here with gnrc_netif_single */
    (void)netdev;
#endif

    eui64_set_local(addr);
    eui64_clear_group(addr);
}
```

Reply:
I think its the same issue described in #13358, on multiple `iotl-lab-m3` they all get the same short-addres `13:15` .

Reply:
> Huh? Byte 7 will change if we got the EUI64 from luid and there are multiple interfaces

I'm taking about uniqueness across platforms, not interfaces.

Reply:
@benpicco do you have an issue with the proposed change?

```c
static inline void eui_short_from_eui64(eui64_t *addr_long,
                                        network_uint16_t *addr_short)
{
    /* https://tools.ietf.org/html/rfc4944#section-12 requires the first bit to
     * 0 for unicast addresses */
    addr_short->u8[0] = addr_long->uint8[0] & 0x7F;
    addr_short->u8[1] ^= addr_long->uint8[1];
}
```

Reply:
Yes! Because on platforms with two interfaces that got their EUI from luid, now both short addresses will be the same.

How is using the first two bytes of a random number going to make them any more unique then using the last two bytes?

Reply:
Because the LONG hardware address last two bits is always the same for `iotlab-m3`

Reply:
See https://github.com/RIOT-OS/RIOT/issues/13358#issuecomment-585657477

Reply:
Uh but that's by chance - I'd rather change `luid_base()` to use a more hash-like algorithm. 
The [`iotlab-m3`](https://github.com/iot-lab/iot-lab/wiki/Hardware-M3-node) has a NOR flash - is there a chance to store a persistent configuration there?

Reply:
> is there a chance to store a persistent configuration there?

What kind of configuration?

> I'd rather change luid_base() to use a more hash-like algorithm.

I agree that would be best, but hasn't this discussion already been had before, I haven't been involved as much in this whole `luid` discussion process.

Reply:
> What kind of configuration?

We could store a proper EUI-64 there (and even a unique short address)

> I agree that would be best, but hasn't this discussion already been had before

Not much discussion of the composition of `luid_base()` so far :wink: 

@maribu did some [experiments](https://github.com/maribu/luid_playground). There the current algorithm doesn't look too bad if I interpret the results correctly.

Reply:
@aabadie do you have some input on the m3? Could we do this?

Reply:
I haven't followed all the discussion in detail here. What I see is that there's a regression introduced on iotlab-m3. Why is it not possible to compute the EUI-64 from cpuid and eventually the interface ?

Reply:
> Why is it not possible to compute the EUI-64 from cpuid and eventually the interface ?

This is what happens, but it seems like the last bytes of the generated EUI-64 are not very rich in entropy.

Reply:
> This is what happens, but it seems like the last bytes of the generated EUI-64 are not very rich in entropy.

Then why the logic from #13362 was changed ?

Reply:
#13362 replaced the logic in the driver with a common helper function.
That it 'solved' this problem was by accident.

The root cause is that the CPU ID (especially the uniform parts) is reflected too directly in the generated ID.
How about we shuffle the bytes a bit?